core/state_processor.go

```diff
@@ -84,8 +84,16 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 		allLogs = append(allLogs, receipt.Logs...)
 	}
 	// Finalize the block, applying any consensus engine specific extras (e.g. block rewards)
+	statedb.Prepare(common.Hash{}, block.Hash(), len(block.Transactions()))
```

Comment:
Why is this needed?

Reply:
When logs are added, the associated transaction will be read from the instance variables that are set by this method.